src/pkcs11.cpp

```diff
@@ -139,6 +139,87 @@ static int pkcs11_initialize(const lmp_options &opt)
 	return 0;
 }
 
+/* Look for keys matching a label, reporting an error or removing them */
+static int check_keys_list(PKCS11_KEY *key, unsigned int n, string label,
+                                                             bool remove)
+{
+	for ( ; key && n; key++, n--) {
+		if (key->label != label)
+			continue;
+		const string type = key->isPrivate? "private" : "public";
+		if (remove) {
+			cout << "Removing existing " << type << " key with \""
+			     << label << "\" label" << endl;
+			if (PKCS11_remove_key(key))
+				leave;
+		} else {
+			cerr << "Found " << type << " key with conflicting \""
+			     << label << "\" label. Aborting." << endl;
+			cerr << "Re-run with --force 1 to cleanup existing "
+				"conflicting keys" << endl;
+			return -1;
+		}
+	}
+	return 0;
+}
+
+/*
+ * Verify if the HSM has public or private keys with the given label.
+ * Abort execution or remove them depending on the remove option
+ */
+static int locate_keys(PKCS11_TOKEN *token, string label, bool remove)
+{
+	PKCS11_KEY *key = NULL;
+	unsigned int n = 0;
+
+	if (PKCS11_enumerate_public_keys(token, &key, &n) || !n)
+		return 0;
+
+	if (check_keys_list(key, n, label, remove))
+		return -1;
+
+	if (PKCS11_enumerate_keys(token, &key, &n) || !n)
+		leave;
+
+	if (check_keys_list(key, n, label, remove))
+		return -1;
+
+	return 0;
+}
+
+/*
+ * Verify if the HSM has certificates with the given label.
+ * Abort execution or remove them depending on the remove option
+ */
+static int locate_certs(PKCS11_TOKEN *token, string label, bool remove)
```

Comment:
instead of locate_certs with an option to remove could we do it in two steps
something like

```
if (certificate_exists(...))
   if (force)
        delete_certificate()
   else
        return -1
```

Reply:
In that case, if `force=true`, we would have to iterate over the certificates twice, since there might be more that 1 certificate with the conflicting label. Should I change it anyway?

Reply:
yes please. I am not too concerned about performance during registration (a few miliseconds more when we are going to the cloud doesnt really matter IMO).

Also could you do the changes in **pkcs11_check_hsm()** instead

Reply:
also notice that if the token is not found there is no point on checking for the presence of keys or certificates

Reply:
> instead of locate_certs with an option to remove could we do it in two steps

> Also could you do the changes in pkcs11_check_hsm() instead

> also notice that if the token is not found there is no point on checking for the presence of keys or certificates

Handled the 3 suggestions above. (Missed the `pkcs11_check_hsm` one initially, so I had to made a second force push.)